core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java

```diff
@@ -411,21 +778,88 @@ public boolean equals(Object obj)
             return false;
         }
         SortedRangeSet other = (SortedRangeSet) obj;
-        return Objects.equals(this.lowIndexedRanges, other.lowIndexedRanges);
+        return hashCode() == other.hashCode() && // compare hash codes because they are cached, so this is cheap and efficient
+                Objects.equals(this.type, other.type) &&
+                Arrays.equals(inclusive, inclusive) &&
+                blocksEqual(this.sortedRanges, other.sortedRanges);
+    }
+
+    private boolean blocksEqual(Block leftBlock, Block rightBlock)
+    {
+        if (leftBlock.getPositionCount() != rightBlock.getPositionCount()) {
+            return false;
+        }
+        for (int position = 0; position < leftBlock.getPositionCount(); position++) {
+            if (!valuesEqual(leftBlock, position, rightBlock, position)) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private boolean valuesEqual(Block leftBlock, int leftPosition, Block rightBlock, int rightPosition)
+    {
+        boolean leftIsNull = leftBlock.isNull(leftPosition);
+        boolean rightIsNull = rightBlock.isNull(rightPosition);
+        if (leftIsNull || rightIsNull) {
+            // TODO this should probably use IS NOT DISTINCT FROM
+            return leftIsNull == rightIsNull;
+        }
+        Boolean equal;
+        try {
+            equal = (Boolean) equalOperator.invokeExact(leftBlock, leftPosition, rightBlock, rightPosition);
```

Comment:
Is it possible that this returns `null`? If not, then please change `B` to `b`

Reply:
i think it can (also, the equals method signature is *B*oolean)